File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.py
def solution(cap, n, deliveries, pickups):
    d_end = len(deliveries) - 1
    p_end = len(pickups) - 1
    
    answer = 0
    
    while d_end >= 0 or p_end >= 0:
        round_d = 0
        round_p = 0
        
        while d_end >= 0 and deliveries[d_end] == 0:
            d_end -= 1
        
        move = d_end+1
        
        while p_end >= 0 and pickups[p_end] == 0:
            p_end -= 1
        
        answer += max(move, p_end+1) * 2
        
        while d_end >= 0 and round_d < cap:
            if round_d + deliveries[d_end] <= cap:
                round_d += deliveries[d_end]
                deliveries[d_end] = 0
                d_end -= 1
            else:
                deliveries[d_end] -= cap-round_d
                round_d += cap - round_d
                
        while p_end >= 0 and round_p < cap:
            if round_p + pickups[p_end] <= cap:
                round_p += pickups[p_end]
                pickups[p_end] = 0
                p_end -= 1
            else:
                pickups[p_end] -= cap-round_p
                round_p += cap - round_p
            
    return answer
                
    
    

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java
class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        int d = n - 1;
        int p = n - 1;
        
        while (d >= 0 && deliveries[d] == 0) {
            d--;
        }
        
        while (p >= 0 && pickups[p] == 0) {
            p--;
        }
        
        long answer = 0;
        
        while (d >= 0 || p >= 0) {
            int start_d = d;
            int start_p = p;
            
            answer += ((int) Math.max(start_d, start_p) + 1) * 2;
            
            int c = cap;
            while (d >= 0 && (c > 0 || deliveries[d] == 0)) {
                if (deliveries[d] > c) {
                    deliveries[d] -= c;
                    c = 0;
                }
                else {
                    c -= deliveries[d];
                    deliveries[d] = 0;
                    d -= 1;
                }
            }
            
            c = cap;
            while (p >= 0 && (c > 0 || pickups[p] == 0)) {
                if (pickups[p] > c) {
                    pickups[p] -= c;
                    c = 0;
                }
                else {
                    c -= pickups[p];
                    pickups[p] = 0;
                    p -= 1;
                }
            }
        }
        return answer;
    }
}